util/reflect.go

```diff
@@ -367,61 +413,368 @@ type NodeInfo struct {
-func ForEachField(value interface{}, in, out interface{}, iterFunction FieldIteratorFunc) (errs []error) {
+func ForEachField(schema *yang.Entry, value interface{}, in, out interface{}, iterFunction FieldIteratorFunc) (errs Errors) {
 	if IsValueNil(value) {
 		return nil
 	}
-	return forEachFieldInternal(&NodeInfo{FieldValue: reflect.ValueOf(value)}, in, out, iterFunction)
+	return forEachFieldInternal(&NodeInfo{Schema: schema, FieldValue: reflect.ValueOf(value)}, in, out, iterFunction)
 }
 
 // forEachFieldInternal recursively iterates through the fields of value (which
 // may be any Go type) and executes iterFunction on each field.
 //   in, out are passed through from the caller to the iteration and can be used
 //     arbitrarily in the iteration function to carry state and results.
-func forEachFieldInternal(ni *NodeInfo, in, out interface{}, iterFunction FieldIteratorFunc) (errs []error) {
-	if IsNilOrInvalidValue(ni.FieldValue) {
+func forEachFieldInternal(ni *NodeInfo, in, out interface{}, iterFunction FieldIteratorFunc) (errs Errors) {
+	if IsValueNil(ni) {
 		return nil
 	}
 
 	errs = AppendErrs(errs, iterFunction(ni, in, out))
 
+	v := ni.FieldValue
+	t := v.Type()
+
 	switch {
-	case IsValueStruct(ni.FieldValue) || IsValueStructPtr(ni.FieldValue):
-		structElems := derefIfStructPtr(ni.FieldValue)
-		for i := 0; i < structElems.NumField(); i++ {
-			nn := *ni
-			nn.ParentStruct = ni.FieldValue.Interface()
-			nn.FieldType = structElems.Type().Field(i)
-			nn.FieldValue = structElems.Field(i)
-			errs = AppendErrs(errs, forEachFieldInternal(&nn, in, out, iterFunction))
+	case IsTypeStructPtr(t):
+		t = t.Elem()
+		if !IsNilOrInvalidValue(v) {
+			v = v.Elem()
+		}
+		fallthrough
+	case IsTypeStruct(t):
+		for i := 0; i < t.NumField(); i++ {
+			sf := t.Field(i)
+			nn := &NodeInfo{
+				Parent:      ni,
+				StructField: sf,
+				FieldValue:  reflect.Zero(sf.Type),
+			}
+			if !IsNilOrInvalidValue(v) {
+				nn.FieldValue = v.Field(i)
+			}
+			ps, err := SchemaPaths(nn.StructField)
+			if err != nil {
+				return NewErrs(err)
+			}
+			for _, p := range ps {
+				nn.Schema = ChildSchema(ni.Schema, p)
+				if nn.Schema == nil {
+					e := fmt.Errorf("forEachFieldInternal could not find child schema with path %v from schema name %s", p, ni.Schema.Name)
+					DbgPrint(e.Error())
+					return NewErrs(e)
+				}
+				nn.PathFromParent = p
+				// In the case of a map/slice, the path is of the form
+				// "container/element".
+				if IsTypeSlice(sf.Type) || IsTypeMap(sf.Type) {
+					nn.PathFromParent = p[0:1]
+				}
+				errs = AppendErrs(errs, forEachFieldInternal(nn, in, out, iterFunction))
+			}
 		}
 
-	case IsValueSlice(ni.FieldValue):
-		for i := 0; i < ni.FieldValue.Len(); i++ {
-			nn := *ni
-			nn.FieldValue = ni.FieldValue.Index(i)
-			errs = AppendErrs(errs, forEachFieldInternal(&nn, in, out, iterFunction))
+	case IsTypeSlice(t):
+		schema := *(ni.Schema)
```

Comment:
Consider doing this outside of the `switch` given that it's used in multiple `case` statements.